ambry-network/src/test/java/com.github.ambry.network/NetworkClientTest.java

```diff
@@ -201,6 +201,46 @@ public void testExceptionOnConnect() {
   }
 
   /**
+   * Test to ensure two things:
+   * 1. If a request comes in and there are no available connections to the destination,
+   * only one connection is initiated, even if that connection is found to be pending when the
+   * same request is looked at during a subsequent sendAndPoll.
+   * 2. For the above situation, if the subsequent pending connection fails, then the request is
+   * immediately failed.
+   */
+  @Test
+  public void testConnectionInitiationFailures()
+      throws Exception {
+    List<RequestInfo> requestInfoList = new ArrayList<>();
+    requestInfoList.add(new RequestInfo(host2, port2, new MockSend(3)));
+    selector.setState(MockSelectorState.IdlePoll);
+    Assert.assertEquals(0, selector.connectCallCount());
+    // this sendAndPoll() should initiate a connect().
+    List<ResponseInfo> responseInfoList = networkClient.sendAndPoll(requestInfoList, 100);
+    // At this time a single connection would have been initiated for the above request.
+    Assert.assertEquals(1, selector.connectCallCount());
+    Assert.assertEquals(0, responseInfoList.size());
+    requestInfoList.clear();
+
```

Comment:
just for comprehensiveness. Can you request another request for the same destination (different request) and it should have triggered a connection request. It shouldn't depend on other connection request. Since we don't know the implementation, there are chances that this could happen